dialogue-core/src/main/java/com/palantir/dialogue/core/ContentDecodingChannel.java

```diff
@@ -0,0 +1,236 @@
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.zip.GZIPInputStream;
+
+/**
+ * Adds support for transparently requesting and decoding <code>Content-Encoding: gzip</code> responses
+ * in a client agnostic way. Client implementations may choose to decompress GZIP data more efficiently
+ * if possible.
+ * This allows client implementations to avoid considering content-encoding in most cases, and
+ * sets a specific <code>Accept-Encoding</code> header to avoid potentially using an unexpected
+ * type based on client defaults (for example apache httpclient requests gzip and deflate by default).
+ * https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.3
+ * https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.11
+ */
+final class ContentDecodingChannel implements Channel {
+
+    private static final String ACCEPT_ENCODING = "accept-encoding";
+    private static final String CONTENT_ENCODING = "content-encoding";
+    private static final String CONTENT_LENGTH = "content-length";
+    private static final String GZIP = "gzip";
+
+    private final Channel delegate;
+
+    ContentDecodingChannel(Channel delegate) {
+        this.delegate = Preconditions.checkNotNull(delegate, "Channel is required");
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+        return Futures.transform(
+                delegate.execute(endpoint, acceptEncoding(request)),
+                ContentDecodingChannel::decompress,
+                MoreExecutors.directExecutor());
+    }
+
+    private static Request acceptEncoding(Request request) {
+        if (request.headerParams().containsKey(ACCEPT_ENCODING)) {
+            // Do not replace existing accept-encoding values
+            return request;
+        }
+        return Request.builder()
+                .from(request)
+                .putHeaderParams(ACCEPT_ENCODING, GZIP)
+                .build();
+    }
+
+    private static Response decompress(Response input) {
+        Optional<String> contentEncoding = input.getFirstHeader(CONTENT_ENCODING);
+        if (contentEncoding.isPresent() && GZIP.equals(contentEncoding.get())) {
+            return new ContentDecodingResponse(input);
+        }
+        return input;
```

Comment:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding#Syntax shows a slightly funky response that servers can apparently send: `Content-Encoding: gzip, identity`... do you think it's worth having a debug log in here in case we ever want to look into why this isn't running?

Reply:
Good call on debug logging, I've never seen multiple Content-Encoding values in practice, but it would be helpful to have logging in this case.